temporalcli/internal/printer/printer.go

```diff
@@ -0,0 +1,514 @@
+	AppendJSONNewline            bool
+}
+
+type Align int
+
+const (
+	AlignDefault Align = tablewriter.ALIGN_DEFAULT
+	AlignCenter        = tablewriter.ALIGN_CENTER
+	AlignRight         = tablewriter.ALIGN_RIGHT
+	AlignLeft          = tablewriter.ALIGN_LEFT
+)
+
+type TableOptions struct {
+	// If not set for a field, maximum width of all rows for structured, and no
+	// width for streaming table. Field width will always at least be field name.
+	FieldWidths map[string]int
+	// Fields are align-left by default
+	FieldAlign map[string]Align
+	NoHeader   bool
+}
+
+// For JSON, if v is a proto message, protojson encoding is used
+func (p *Printer) PrintStructured(v any, options StructuredOptions) error {
+	// JSON
+	if p.JSON {
+		return p.printJSON(v, options)
+	}
+
+	// Get data
+	cols := options.toPredefinedCols()
+	cols, rows, err := p.tableData(cols, v)
+	if err != nil {
+		return err
+	}
+	cols = adjustColsToOptions(cols, options)
+
+	// Text table
+	if options.Table != nil {
+		p.calculateUnsetColWidths(cols, rows)
+		p.printTable(options.Table, cols, rows)
+		return nil
+	}
+
+	// Text "card"
+	p.printCards(cols, rows)
+	return nil
+}
+
+type PrintStructuredIter interface {
+	// Nil when done
+	Next() (any, error)
+}
+
+// Fields must be present for table
+func (p *Printer) PrintStructuredIter(typ reflect.Type, iter PrintStructuredIter, options StructuredOptions) error {
+	cols := options.toPredefinedCols()
+	if !p.JSON {
+		if len(cols) == 0 {
+			var err error
+			if cols, err = deriveCols(typ); err != nil {
+				return fmt.Errorf("unable to derive columns: %w", err)
+			}
+		}
+		cols = adjustColsToOptions(cols, options)
+		// We're intentionally not calculating field lengths and only accepting them
+		// since this is streaming
+		if options.Table != nil {
+			p.printHeader(cols)
+		}
+	}
+	for {
+		v, err := iter.Next()
+		if v == nil || err != nil {
+			return err
+		}
+		if p.JSON {
+			b, err := json.Marshal(v)
+			if err != nil {
+				return err
+			}
+			p.write(b)
+			p.writeStr("\n")
+		} else {
+			row, err := p.tableRowData(cols, v)
+			if err != nil {
+				return err
+			}
+			if options.Table != nil {
+				p.printRow(cols, row)
+			} else {
+				p.printCard(cols, row)
+			}
+		}
+	}
+}
+
+func (p *Printer) write(b []byte) {
+	if _, err := p.Output.Write(b); err != nil {
+		panic(err)
+	}
+}
+
+func (p *Printer) writeStr(s string) {
+	p.write([]byte(s))
+}
+
+func (p *Printer) writef(s string, v ...any) {
+	if _, err := fmt.Fprintf(p.Output, s, v...); err != nil {
+		panic(err)
+	}
+}
+
+func (p *Printer) printJSON(v any, options StructuredOptions) error {
+	shorthandPayloads := p.JSONPayloadShorthand
+	if options.OverrideJSONPayloadShorthand != nil {
+		shorthandPayloads = *options.OverrideJSONPayloadShorthand
+	}
+	b, err := p.jsonVal(v, p.JSONIndent, shorthandPayloads)
+	if err != nil {
+		return err
+	}
+	_, err = p.Output.Write(b)
+	if err == nil && options.AppendJSONNewline {
+		_, err = p.Output.Write([]byte("\n"))
```

Comment:
Seems unnecessary when just dumping JSON to stdout and only necessary when dumping adjacent JSON for streaming. But I agree, just because it's unnecessary doesn't mean it's a problem. I will update.

Reply:
Updated